### PR TITLE
Expose more information in `get_body_with_borrowck_facts`

### PR DESCRIPTION
**Pull request by voidc**
Verification tools for Rust such as, for example, Creusot or Prusti would benefit from having access to more information computed by the borrow checker.
As a first step in that direction, #86977 added the `get_body_with_borrowck_facts` API, allowing compiler consumers to obtain a `mir::Body` with accompanying borrow checker information.
At RustVerify 2023, multiple people working on verification tools expressed their need for a more comprehensive API.
While eventually borrow information could be part of Stable MIR, in the meantime, this PR proposes a more limited approach, extending the existing `get_body_with_borrowck_facts` API.
In summary, we propose the following changes:

- Permit obtaining the borrow-checked body without necessarily running Polonius
- Return the `BorrowSet` and the `RegionInferenceContext` in `BodyWithBorrowckFacts`
- Provide a way to compute the `borrows_out_of_scope_at_location` map
- Make some helper methods public

This is similar to #108328 but smaller in scope.
@smoelius Do you think these changes would also be sufficient for your needs?

r? @oli-obk 
cc @JonasAlaif

### CODE REVIEW

File: compiler/rustc_borrowck/src/lib.rs [hunk 1]
```diff
@@ -469,12 +470,20 @@ fn do_mir_borrowck<'tcx>(
...
-    pub output_facts: Rc<self::nll::PoloniusOutput>,
-    /// The table that maps Polonius points to locations in the table.
-    pub location_table: LocationTable,
+    /// The set of borrows occurring in `body` with data about them.
+    pub borrow_set: Rc<BorrowSet<'tcx>>,
+    /// Context generated during borrowck, intended to be passed to
+    /// [`OutOfScopePrecomputer`](dataflow::OutOfScopePrecomputer).
+    pub region_inference_context: Rc<RegionInferenceContext<'tcx>>,
+    /// The table that maps Polonius points to locations in the table. Populated
+    /// when using [`ConsumerOptions::PoloniusInputFacts`] or above.
```
Comment (oli-obk) [inline compiler/rustc_borrowck/src/lib.rs:482]: what does "or above" mean?

File: compiler/rustc_borrowck/src/lib.rs [hunk 2]
```diff
@@ -444,13 +443,15 @@ fn do_mir_borrowck<'tcx>(
...
-        let output_facts = mbcx.polonius_output.expect("Polonius output was not computed");
+    let body_with_facts = if consumer_options.is_some() {
+        let output_facts = mbcx.polonius_output;
         Some(Box::new(BodyWithBorrowckFacts {
             body: body_owned,
-            input_facts: *polonius_input.expect("Polonius input facts were not generated"),
+            borrow_set,
+            region_inference_context: regioncx,
+            location_table: polonius_input.as_ref().map(|_| location_table_owned),
+            input_facts: polonius_input,
```
Comment (oli-obk) [inline compiler/rustc_borrowck/src/lib.rs:454]: These two fields are always `Some` or `None` together, it may make sense to merge them into a common struct and put that into the `Option`

File: compiler/rustc_borrowck/src/lib.rs [hunk 3]
```diff
@@ -470,6 +471,8 @@ fn do_mir_borrowck<'tcx>(
 pub struct BodyWithBorrowckFacts<'tcx> {
     /// A mir body that contains region identifiers.
     pub body: Body<'tcx>,
+    /// The mir bodies of promoteds.
+    pub promoted: IndexVec<Promoted, Body<'tcx>>,
```
Comment (oli-obk) [inline compiler/rustc_borrowck/src/lib.rs:475]: Note that these may disappear soon: https://github.com/rust-lang/rust/pull/111693

**Review by oli-obk — CHANGES_REQUESTED** [review]
These changes look reasonable to me. I cannot promise we'll keep the API around, but I do not foresee this being a maintenance issue for us, so likely some form of it will keep getting provided.

**Comment by smoelius** [comment]
> @smoelius Do you think these changes would also be sufficient for your needs?

I think it might. This looks very similar to some local changes that I have. Thank you very much for the ping.

File: compiler/rustc_borrowck/src/lib.rs (hunk 1 above)
Comment (voidc) [reply compiler/rustc_borrowck/src/lib.rs]: "or above" refers to the hierarchy of the `ConsumerOptions`. I changed the wording to make that clearer.

File: compiler/rustc_borrowck/src/lib.rs (hunk 2 above)
Comment (voidc) [reply compiler/rustc_borrowck/src/lib.rs:453]: I agree but I can't think of a reasonable name to give to that struct.

File: compiler/rustc_borrowck/src/lib.rs (hunk 3 above)
Comment (voidc) [reply compiler/rustc_borrowck/src/lib.rs]: I've been following that PR as it would also simplify how we consume MIR. Should this PR wait until after #111693 is merged?

File: compiler/rustc_borrowck/src/lib.rs (hunk 2 above)
Comment (oli-obk) [reply compiler/rustc_borrowck/src/lib.rs:453]: yea... it's not clear what's best to me either. Especially since there's the output facts that partially overlap, too.

File: compiler/rustc_borrowck/src/lib.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_borrowck/src/lib.rs]: afaict it will not be an issue for that PR to just remove this field here after a rebase

**Comment by oli-obk** [comment]
please rebase, then we can land this

**Comment by smoelius** [comment]
Sorry if it's not my place to chime in, but will the right commit be pushed? (I can't tell from the [queue](https://bors.rust-lang.org/homu/queue/rust).)

I think @voidec intends for it to be [8dac074](https://github.com/rust-lang/rust/pull/111840/commits/8dac074087afc9f02a794027bede787bb9367ad3).

But that was pushed after @oli-obk [approved](https://github.com/rust-lang/rust/pull/111840#issuecomment-1559198333).

**Comment by oli-obk** [comment]
The moment a change is done to a PR it is taken off the queue, so nothing is happening for now until I actually review and r+ it again

**Comment by voidc** [comment]
Yes, the PR is ready

@rustbot ready

**Comment by lqd** [comment]
> This seems like a small change to be able to get all of the `OutlivesConstraint`s out, which already happen to be public.

Depending on what you want to achieve, I don't believe `OutlivesConstraint`s would be enough to have a somewhat "complete outlives graph" view of things, since they will not contain member constraints whenever opaque types are involved. (They're not separate in the NLL MIR dump, nor visible in the graphviz region and SCC graphs, and neither present in the polonius input facts IIRC)

```rust
pub trait T {}

struct S<'a>(&'a ());

impl<'a> T for S<'a> {}

fn foo() -> impl T {
    let x = ();
    S(&x) //~ ERROR `x` does not live long enough
}
```

```
error[E0597]: `x` does not live long enough
  --> test.rs:9:7
   |
8  |     let x = ();
   |         - binding `x` declared here
9  |     S(&x) //~ ERROR `x` does not live long enough
   |     --^^-
   |     | |
   |     | borrowed value does not live long enough
   |     opaque type requires that `x` is borrowed for `'static`
10 | }
   | - `x` dropped here while still borrowed
```

The `constraints` field won't have an outlives constraint flowing into 'static here.
